fsdp/minimal-fsdp/fsdp.py

```diff
@@ -0,0 +1,339 @@
+        with core_context.checkpoint.store_path(
+            metadata={"steps_completed": steps_completed}
+        ) as (
+            path,
+            _,
+        ):
+            torch.save(model_state_dict, path.joinpath("model.bin"))
+            torch.save(optim_state_dict, path.joinpath("optim.bin"))
+            if use_amp:
+                # Scaler state is automatically the same across ranks.
+                scaler_state_dict = scaler.state_dict()
+                torch.save(scaler_state_dict, path.joinpath("scaler.bin"))
+
+
+def load_checkpoint(
+    fsdp_model: FSDP,
+    optimizer: torch.optim.Optimizer,
+    scaler: ShardedGradScaler,
+    use_amp: bool,
+    core_context: det.core.Context,
+    device: torch.device,
+    uuid: str,
+) -> int:
+    with core_context.checkpoint.restore_path(uuid) as path:
+        with FSDP.state_dict_type(
+            fsdp_model,
+            StateDictType.FULL_STATE_DICT,
+            FullStateDictConfig(offload_to_cpu=True, rank0_only=True),
+        ):
+            fsdp_model.load_state_dict(
+                torch.load(path.joinpath("model.bin"), map_location=device)
+            )
+            optim_state_dict = torch.load(
+                path.joinpath("optim.bin"), map_location=device
+            )
+            optim_state_dict_to_load = FSDP.optim_state_dict_to_load(
+                model=fsdp_model,
+                optim=optimizer,
+                optim_state_dict=optim_state_dict,
+            )
+            optimizer.load_state_dict(optim_state_dict_to_load)
+        scaler_path = path.joinpath("scaler.bin")
+        if use_amp and os.path.isfile(scaler_path):
+            scaler.load_state_dict(torch.load(scaler_path))
+
+        with open(path.joinpath("metadata.json"), "r") as f:
+            metadata = json.load(f)
+
+    last_step_completed = metadata["steps_completed"]
+    return last_step_completed
+
+
+def main(
+    core_context: det.core.Context,
+    hparams: dict[str, Any],
+    checkpoint_uuid: Optional[str] = None,
+) -> None:
+    # Fix the random seed on all devices
+    seed = 42
+    random.seed(seed)
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+
+    # Get and set the device for this process
+    device = torch.device(f"cuda:{core_context.distributed.local_rank}")
+    torch.cuda.set_device(device)
+
+    # Build the unsharded model directly on the device.
+    model = Transformer(
+        d_model=hparams["d_model"],
+        n_heads=hparams["n_heads"],
+        vocab_size=hparams["vocab_size"],
+        n_layers=hparams["n_layers"],
+        max_seq_len=hparams["max_seq_len"],
+        device=device,
+    )
+
+    # Inspect the model:
+    if core_context.distributed.rank == 0:
+        print("Model before FSDP:")
+        print(model, flush=True)
+
+    # Wrap the embedding layer, the lm head, and each transformer block into its own FSDP unit:
+    auto_wrap_policy = ModuleWrapPolicy([TransformerBlock, EmbedAndEncode, LMHead])
+
```

Comment:
I should have caught this before, but there are some small `use_amp = True` related issues we probably want to adress.

First, there is some special handling for the FSDP weight/comms/etc types when using amp, as in
```python
from torch.distributed.fsdp import MixedPrecision
fsdp_model = FSDP(model,
    mixed_precision=MixedPrecision(param_dtype=torch.bfloat16,
    reduce_dtype=torch.bfloat16), ...)
``` 
Not needed for correctness, just efficiency.

Second,  for the autocast in `get_loss` it's better practice to specify the `dtype=torch.bfloat16` arg, as in 
```python
with torch.cuda.amp.autocast(enabled=use_amp, dtype=torch.bfloat16):
        outputs = fsdp_model(inputs)
```
The default is `dtype=torch.float16` (not `bfloat16`).

Third, with `bfloat16` there should be no need for the `ShardedGradScaler`; only needed for `float16`.

Reply:
Apologies, I should have caught these. Think I missed them when just looking at the diffs.

Reply:
> First, there is some special handling for the FSDP weight/comms/etc types when using amp ... Not needed for correctness, just efficiency.

Good catch, wasn't aware of this; I assume `autocast` just fails to recognize / handle some of the buffers and operations used inside of FSDP without it?

> Second, for the autocast in get_loss it's better practice to specify the dtype=torch.bfloat16 arg, as in

`bfloat16` isn't supported for pre-A100 GPUs, right?  I imagine we want our example to run on any GPU, hence defaulting to `float16`.  Doing a quick Google search, looks like this can be made conditional via e.g.
```
compute_capability = torch.cuda.get_device_capability()
if compute_capability[0] < 8:
     ....
```
So we could switch to using `bfloat16` iff it's supported.

> Third, with bfloat16 there should be no need for the ShardedGradScaler; only needed for float16.

See above, but yes -- if we switch to `bfloat16` entirely or conditionally, we could omit.

Reply:
Good call on hardware compatibility. Apparently if the hardware doesn't support `bfloat16`, [it silently falls back to `float32`?](https://discuss.pytorch.org/t/bfloat16-on-nvidia-v100-gpu/201629/4) Seem weird.

Anyway, your call on what to support there. Could also make it configurable as another hparam, `amp_dtype`?

> I assume autocast just fails to recognize / handle some of the buffers and operations used inside of FSDP without it?

Nah, it's a little different, IIUC. The actual weights used for the forwards are held in whatever precision you specify in `MixedPrecision`, or `float32` if omitted. Under autocasting, various tensors have their `dtypes` changed so that specific operations occur in either high or low precision. E.g. matmuls in low-precision and softmax in high-precision.  `MixedPrecision` affects which direction most of the casts occur in (among other things).

Like in non-FSDP weights would always be kept in high-precision and down-cast as needed. With FSDP and `MixedPrecision(param_dtype=torch.bfloat16)`, say, you also avoid those down-casts because the weights are already in the desired precision. But you then might need some extra up-casts elsewhere.

There are no explicit failures if you don't specify `MixedPrecision` w/ FSDP and run under `autocast`; just perf differences, and likely some (hopefully) small numerical differences also.

I likely got some details incorrect here, also. The FSDP + amp API isn't super well documented. Talked to one of the [OLMo](https://github.com/allenai/OLMo) researchers about exactly this topic and how it's confusing a while ago.

Reply:
Alternatively, if you specify `MixedPrecision` in FSDP but don't wrap in `autocast` none of the casts ever happen and you can end up running every part of the forwards in low-precision, rather than mixed.